src/partition/repart.c

```diff
@@ -479,6 +480,43 @@ static uint64_t round_up_size(uint64_t v, uint64_t p) {
         return v * p;
 }
 
+/* calculates the size of a dm-verity hash partition's contents */
+static int calculate_verity_hash_size(
+                uint64_t *ret_bytes,
+                uint64_t data_bytes,
+                uint64_t hash_block_size,
+                uint64_t data_block_size) {
+        uint8_t digest_size = SHA256_DIGEST_SIZE;
+
+        uint64_t data_blocks = DIV_ROUND_UP(data_bytes, data_block_size);
+        if (data_blocks > UINT64_MAX / data_block_size)
+                return -1; /* overflow */
+
+        /* hashes that fit in one hash block (node in the merkle tree) */
+        uint64_t hashes_per_hash_block = hash_block_size / digest_size;
+
+        /* initialize with 1 for the root of the merkle tree */
+        uint64_t hash_blocks = 1;
+
+        /* iterate through the levels of the merkle tree bottom up */
+        uint64_t remaining_blocks = data_blocks;
+        uint64_t hash_blocks_for_level;
```

Comment:
Please move the declaration of `hash_blocks_for_level` into the `while` loop below.